### PR TITLE
O3-1284: Global patient search and patient search extension V2

### PR DESCRIPTION
**Pull request by vasharma05**
## Requirements

- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [x] My work conforms to the [**OpenMRS 3.0 Styleguide**](https://om.rs/styleguide).
- [x] I checked for feature overlap with [**existing widgets**](https://om.rs/directory).

## Summary
This PR brings the new Patient search design and implementation to the OpenMRS 3.
The key features are:
1. Faster search
2. Compact search implemented
3. Separate patient banner with multiple actions for patient search.
4. Separate patient search page
5. Patient search extension. 

**P.S. Multiple tests are failing on the mono-repo, and I'm figuring it out. The work for Patient Search v2 is completed.**

## Screenshots

Adding soon.

## Related Issue

https://issues.openmrs.org/browse/O3-1284

This PR also solves:
https://issues.openmrs.org/browse/O3-777
https://issues.openmrs.org/browse/O3-1140
https://issues.openmrs.org/browse/O3-682

## Other

*None.*

### CODE REVIEW

File: packages/esm-patient-search-app/src/compact-patient-search/compact-patient-search.scss [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  width: 50vw;
+  position: relative;
+}
+
+.floatingSearchResultsContainer {
+  position: absolute;
+  overflow-y: auto;
+  box-shadow: 0 $spacing-03 $spacing-05 $ui-03;
+  z-index: 99;
+  border: 0 0.063rem 0.063rem 0.063rem solid $ui-03;
```
Comment (ZacButko) [inline packages/esm-patient-search-app/src/compact-patient-search/compact-patient-search.scss:13]: is rem better than pixels for border width? this renders to 1px, but I've read that if it calculates to less than a pixel some browsers won't render it?

File: packages/esm-patient-search-app/src/patient-search-page/patient-banner/contact-details/relationships.resource.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
+import { openmrsFetch } from '@openmrs/esm-framework';
+import useSWR from 'swr';
+
+const customRepresentation =
```
Comment (ZacButko) [inline packages/esm-patient-search-app/src/patient-search-page/patient-banner/contact-details/relationships.resource.tsx:4]: This is such a wild API...

File: packages/esm-patient-search-app/src/patient-search-page/patient-search-lg.component.tsx [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            <div className={`${styles.pagination} ${stickyPagination && styles.stickyPagination}`}>
+              <Pagination setPage={setPage} currentPage={currentPage} hasMore={hasMore} totalPages={totalPages} />
+            </div>
+          )}
+        </div>
+      </div>
+    );
+  }
+
+  return !fetchError ? (
```
Comment (ZacButko) [inline packages/esm-patient-search-app/src/patient-search-page/patient-search-lg.component.tsx:95]: These nested ternaries are a little hard to read. Can you reformat this to be more declarative? E.g.
```
if (fetchError) { return ...} 
if (!fetchError && isEmpty(searchResults)) { return ...} 
```

File: packages/esm-patient-search-app/src/patient-search.resource.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      isLoading: !data?.data && !error,
+      fetchError: error,
+      hasMore: data?.data?.links?.some((link) => link.rel === 'next'),
+      loadingNewData: isValidating,
+      totalResults: data?.data?.totalCount,
+    }),
+    [data, isValidating, error],
+  );
+
+  return results;
```
Comment (ZacButko) [inline packages/esm-patient-search-app/src/patient-search.resource.tsx:52]: For these hooks wouldn't you also be interested in returning `isLoading` and `error` also?

File: packages/esm-patient-search-app/src/patient-search.resource.tsx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      hasMore: data?.length ? !!data[data.length - 1].data?.links?.some((link) => link.rel === 'next') : false,
+      loadingNewData: isValidating,
+      setPage: setSize,
+      currentPage: size,
+      totalResults: data?.[0]?.data?.totalCount,
+    }),
+    [data, isValidating, error, setSize, size],
+  );
+
+  return results;
```
Comment (ZacButko) [inline packages/esm-patient-search-app/src/patient-search.resource.tsx:98]: same here

**Review by ZacButko — COMMENTED** [review]
Lots of good work here. Thanks @vasharma05 . Give a shout out when it turns from draft into finalized PR

File: packages/esm-patient-search-app/src/patient-search.resource.tsx (hunk 4 above)
Comment (vasharma05) [reply packages/esm-patient-search-app/src/patient-search.resource.tsx:52]: I have it in the results object, checkout the line 43.
Thanks!

File: packages/esm-patient-search-app/src/patient-search.resource.tsx (hunk 5 above)
Comment (ZacButko) [reply packages/esm-patient-search-app/src/patient-search.resource.tsx:98]: whoops

File: packages/esm-patient-search-app/src/compact-patient-search/compact-patient-banner.component.tsx [hunk 6]
```diff
@@ -39,23 +18,16 @@ const PatientSearchResults: React.FC<PatientSearchResultsProps> = ({ patients, h
   const getGender = (gender) => {
```
Comment (ZacButko) [inline packages/esm-patient-search-app/src/compact-patient-search/compact-patient-banner.component.tsx:18]: Possible candidate for core

**Review by ZacButko — APPROVED** [review]
Looks great!

File: packages/esm-patient-search-app/src/patient-search-page/patient-banner/contact-details/relationships.resource.tsx (hunk 2 above)
Comment (vasharma05) [reply packages/esm-patient-search-app/src/patient-search-page/patient-banner/contact-details/relationships.resource.tsx:4]: Yes it is 😆

Comment (ibacher) [reply packages/esm-patient-search-app/src/patient-search-page/patient-banner/contact-details/relationships.resource.tsx:4]: Relationships really are one of the weirdest bits of modelling in OpenMRS... In a way it's too flexible to actually be valuable (many real-world implementations simply use a patient attribute or a obs to store the actual related-person contact details... so make sure this is swappable or hideable).

Comment (vasharma05) [reply packages/esm-patient-search-app/src/patient-search-page/patient-banner/contact-details/relationships.resource.tsx:4]: Sure @ibacher , but I'll make these changes in the next iteration.
Merging this one at the moment.

**Review by ZacButko — APPROVED** [review]

**Pull request merged**